### PR TITLE
Add more ABI test cases to miri (RFC 3391)

### PR DESCRIPTION
**Pull request by Lokathor**
Part of https://github.com/rust-lang/rust/issues/110503

cc @RalfJung

### CODE REVIEW

**Comment by Lokathor** [comment]
This doesn't touch line 91 so i'm not sure why it shows that line as having a conflict. I updated my fork just moments before adding these test lines so I'd be shocked if it was out of date already.

**Comment by Lokathor** [comment]
Oh no I think this all went wrong.

**Comment by RalfJung** [comment]
Looks like we need changes to the ABI check logic as well, not just a new test.

Currently, we special-case `Option` [here](https://github.com/rust-lang/rust/blob/e8379c9598d703ec26b967e9654d016c21497beb/compiler/rustc_const_eval/src/interpret/terminator.rs#L296), so some logic will have to be added there to also detect `Result` where one side is a 1-ZST.

**Comment by Lokathor** [comment]
@saethlin said (via community discord) they would be able to add the additional logic some time soon.

**Comment by saethlin** [comment]
> r? @saethlin perhaps then?

Au contraire, I'm implementing this so I do not think I am a suitable reviewer.

File: compiler/rustc_const_eval/src/interpret/terminator.rs [hunk 1]
```diff
@@ -294,17 +294,30 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
...
-        let inner = match layout.ty.kind() {
-            ty::Adt(def, args) if self.tcx.is_diagnostic_item(sym::Option, def.did()) => {
-                args[0].as_type().unwrap()
-            }
-            _ => {
-                // Not an `Option`.
+        // Check if this is `Option` wrapping some type or if this is `Result` wrapping a 1-ZST and
+        // another type.
+        let ty::Adt(def, args) = layout.ty.kind() else {
+            // Not an `Option` or Result.
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/terminator.rs:300]:
```suggestion
            // Not an ADT, so definitely no NPO.
```

File: compiler/rustc_const_eval/src/interpret/terminator.rs [hunk 2]
```diff
@@ -294,17 +294,30 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
...
+        } else if self.tcx.is_diagnostic_item(sym::Result, def.did()) {
+            // We want to extract which (if any) of the args is not a 1-ZST.
+            let lhs = self.layout_of(args[0].as_type().unwrap())?;
+            let rhs = self.layout_of(args[1].as_type().unwrap())?;
+            if lhs.is_1zst() {
+                rhs
+            } else if rhs.is_1zst() {
+                lhs
+            } else {
                 return Ok(layout);
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/terminator.rs:315]:
```suggestion
                return Ok(layout); // no NPO
```

File: compiler/rustc_const_eval/src/interpret/terminator.rs [hunk 3]
```diff
@@ -294,17 +294,30 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
...
             }
+        } else {
+            return Ok(layout);
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/terminator.rs:318]:
```suggestion
            return Ok(layout); // no NPO
```

**Review by RalfJung — COMMENTED** [review]
LGTM modulo some comments. r=me after resolving those.

**Comment by fmease** [comment]
bors sleepy @bors r-

**Comment by Lokathor** [comment]
So, wait, did it actually get merged or not?

**Comment by RalfJung** [comment]
Yeah that is done by github, not bors.

**Pull request merged**